plugins/BEdita/Core/tests/TestCase/Model/Entity/DateRangeTest.php

```diff
@@ -0,0 +1,652 @@
+
+        static::assertSame($expected, $result);
+    }
+
+    /**
+     * Data provider for `testNormalize` test case.
+     *
+     * @return array
+     */
+    public function normalizeProvider()
+    {
+        return [
+            'empty' => [
+                [],
+                [],
+            ],
+            'sort' => [
+                [
+                    [
+                        'start_date' => '2017-01-01',
+                        'end_date' => '2017-01-10',
+                    ],
+                    [
+                        'start_date' => '2017-01-20',
+                        'end_date' => null,
+                    ],
+                    [
+                        'start_date' => '2017-02-01',
+                        'end_date' => '2017-02-10',
+                    ],
+                ],
+                [
+                    [
+                        'start_date' => '2017-02-01',
+                        'end_date' => '2017-02-10',
+                    ],
+                    [
+                        'start_date' => '2017-01-01',
+                        'end_date' => '2017-01-10',
+                    ],
+                    [
+                        'start_date' => '2017-01-20',
+                        'end_date' => null,
+                    ],
+                ],
+            ],
+            'merge' => [
+                [
+                    [
+                        'start_date' => '2017-01-01',
+                        'end_date' => '2017-01-10',
+                    ],
+                    [
+                        'start_date' => '2017-01-20',
+                        'end_date' => null,
+                    ],
+                    [
+                        'start_date' => '2017-02-01',
+                        'end_date' => '2017-02-10',
+                    ],
+                ],
+                [
+                    [
+                        'start_date' => '2017-02-01',
+                        'end_date' => '2017-02-07',
+                    ],
+                    [
+                        'start_date' => '2017-01-01',
+                        'end_date' => '2017-01-05',
+                    ],
+                    [
+                        'start_date' => '2017-02-02',
+                        'end_date' => '2017-02-06',
+                    ],
+                    [
+                        'start_date' => '2017-02-04',
+                        'end_date' => null,
+                    ],
+                    [
+                        'start_date' => '2017-01-05',
+                        'end_date' => '2017-01-10',
+                    ],
+                    [
+                        'start_date' => '2017-01-20',
+                        'end_date' => null,
+                    ],
+                    [
+                        'start_date' => '2017-02-07',
+                        'end_date' => '2017-02-10',
+                    ],
+                ],
+            ],
```

Comment:
extra tests including malformed data (as in previous comments)? it can be tricky, but useful to understand what happens when data ranges are not well formed (error? data is ignored? etc.)